### ISSUE TITLE
Angular material local linking with yarn

### ISSUE THREAD

**Issue by IlCallo** [angular/material2#14270]
#### Bug, feature request, or proposal:
Make yarn local linking work with Angular Material

#### What is the expected behavior?
Using `yarn link` inside a folder of Angular Material monorepo should register correctly for local linking.
Then using `yarn link @angular/<cdk/material/...>` should correctly link the dependency and `ng serve` should work without any errors.

#### What is the current behavior?
Entering single folders and using `yarn link` links the folder itself.
This is a problem because the packages released **ARE NOT** the plain folder, but are generated with `gulp build-aot:release-packages` (or single commands `gulp <material/cdk/...>:build-release`).

Running that command and manually copying the generated folders from `dist/releases` folder will do the job but it's a pretty awful solution.

There are a lot of gulp tasks not mentioned inside DEV_ENVIRONMENT and which should probably be.
For example `gulp cdk:build-release`, which is the CDK counterpart of the material task, and `gulp build-aot:release-packages` which builds all packages (material, cdk, cdk-experimental, material-experimental, material-moment-adapter and material-examples).

It's not possible to link dist folders with yarn, because there are no package.json files into the distribution from which to take the package name (or, at least, I think that's the reason)

#### What is the use-case or motivation for changing an existing behavior?
Enabling a simpler way to test in local environment will help to try out the changes directly on complex projects where components are used, instead of trying a PR and waiting until following release or relying only on the ad-hoc and simpler examples inside the repo.

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
Angular 7.x
Material 7.x

**Comment by crisbeto** [comment]
For what it's worth, we publish each commit that is merged to https://github.com/angular/material2-builds which you can use to try out particular changes.

**Comment by IlCallo** [comment]
That's a useful info, thanks.
But that can be used only after the merge and this can literally take ages.

Example: last month I contributed with stepper label implementation of positioning. In examples it worked ok, then when it was released I found out that in real case scenario it had a bug (really simple to spot, except that on examples page).

Now I'm working on fixing nested table sticky rows on Edge, I made it work on example page, then I copied release packages onto my real world project and some basic things don't work.
And nobody is actually answering to the PR I made (this is not a critic, just a fact) so I can't even hope to get it merged at some point.

Only way for me is working and trying it locally, by copy/paste atm.
I don't know if it's possible to define a different folder target for `yarn link` in package.json, I'm actually searching for that kind of option to make linking easier

**Comment by IlCallo** [comment]
> It's not possible to link dist folders with yarn, because there are no package.json files into the distribution from which to take the package name (or, at least, I think that's the reason)

>  If you run the release gulp tasks, there should be package.json files in the release output. So, you should be able to do yarn link inside of those?

Seems like I was wrong on this, I tried to link them from dist folder some days ago and didn't work, maybe I did something wrong, I'll retry tomorrow, ty!

Maybe just adding this kind of information to DEV_ENVIRONMENT file can be enough for now, so other people don't have to manually look into the code to understand how to generate and link a package locally.

The whole point of this issue was actually to document how to do that for other people/future me, so I'm already happy to have an answer now :P

**Comment by IlCallo** [comment]
I tryed linking from inside release folders and linking succeeded.
But then, when compiling, this error fired off for every single module imported from cdk or material packages:

`Module not found: Error: Can't resolve '@angular/material/<module>' in 'C:\...\material2\dist\releases\material\esm5'
ERROR in C:/.../material2/dist/releases/material/esm5/material.es5.js`

Ideas on what it could be?

**Comment by DevVersion** [comment]
Hmm, I can't think of any reason why this doesn't work. I can try tomorrow and see if it works for me.

Are you just using a CLI project?

**Comment by DevVersion** [comment]
@IlCallo Sorry for the late response. I've looked into it and even tried linking the Github snapshot build repositories, but those didn't work either.

Since they work if you just add them like a normal package, but fail with `yarn link`, I feel like Yarn's link feature does not properly handle multiple `package.json` files in a linked package (or secondary entry-points).

As I think there is nothing we could do about and it would never become a high priority, I will be closing that issue. Happy to chat more about it here though.

**Comment by IlCallo** [comment]
I've opened a new issue on yarn repository for this, let's see if they can help us out...
